README.md
# SafetyNetAttestation

A Ruby gem to verify SafetyNet attestation statements from Google Play Services on your server.

This gem verifies that the statement:
- has a valid signature that is trusted using certificates from https://pki.goog/
- has the correct nonce
- has been generated recently (default allowed leeway from current time is 60 seconds)
- has a signing certificate with the correct subject

With a valid statement your application can then inspect the information contained about the device integrity, calling
app, and if applicable any integrity errors and potential solutions (see usage).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'safety_net_attestation'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install safety_net_attestation

## Usage

Request an attestation statement as described in the [Android developer documentation](https://developer.android.com/training/safetynet/attestation#request-attestation-process) and send the JWS response to your server application.

In your server application code, do the following:

```ruby
require "safety_net_attestation"

statement = begin 
  SafetyNetAttestation::Statement.new(jws_response).verify(nonce)
rescue SafetyNetAttestation::Error => e
  # Statement is not valid, you should abort
end

statement.json
# => {"apkPackageName": "com.package.name.of.requesting.app", "ctsProfileMatch": true, ... }

# snake cased convenience methods are available after #verify call succeeded, use these to make your specific checks: 
statement.cts_profile_match?
# => true
statement.basic_integrity?
# => true
statement.apk_package_name
# => "com.package.name.of.requesting.app"
statement.apk_certificate_digest_sha256
# => ["..."]
statement.error
# => nil
statement.advice
# => nil
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `bin/rspec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/bdewater/safety_net_attestation. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

The gem and its authors are unaffiliated with Google.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.0] - 2019-12-29
### Fixed
- Root certificate loading when this gem is used as a dependency

### Changed
- Rename `Statement#certificates` to `Statement#certificate_chain`

## [0.3.0] - 2019-12-29
### Added
- `Statement#certificates` exposes the certificate chain used during verification
- `Statement#verify` takes an optional `time` argument, defaulting to the current time. This can be used for testing
  captured statements from real devices without needing stubbing.

## [0.2.0] - 2019-12-28
### Fixed
- Fixed loading bundled root certificates

## [0.1.0] - 2019-12-28
### Added
- Extracted from [webauthn-ruby](https://github.com/cedarcode/webauthn-ruby) after discussion with the maintainers. Thanks for the feedback @grzuy and @brauliomartinezlm!

[Unreleased]: https://github.com/bdewater/safety_net_attestation/compare/v0.1.0...HEAD
[0.4.0]: https://github.com/bdewater/safety_net_attestation/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/bdewater/safety_net_attestation/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/bdewater/safety_net_attestation/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/bdewater/safety_net_attestation/releases/tag/v0.1.0


spec/spec_helper.rb
# frozen_string_literal: true

require "bundler/setup"
require "safety_net_attestation"

require "pry-byebug"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.order = :random
end

def generate_key
  OpenSSL::PKey::RSA.new(2048)
end

# rubocop:disable Naming/UncommunicativeMethodParamName
def generate_cert(dn, key, serial, issuer, not_before: nil, not_after: nil)
  cert = OpenSSL::X509::Certificate.new
  issuer = cert unless issuer
  cert.version = 2
  cert.serial = serial
  cert.subject = dn
  cert.issuer = issuer.subject
  cert.public_key = key
  now = Time.now
  cert.not_before = not_before || now - 3600
  cert.not_after = not_after || now + 3600
  cert
end

def issue_cert(dn, key, serial, extensions, issuer: nil, issuer_key: nil, not_before: nil, not_after: nil)
  cert = generate_cert(dn, key, serial, issuer, not_before: not_before, not_after: not_after)
  issuer = cert unless issuer
  issuer_key = key unless issuer_key
  ef = OpenSSL::X509::ExtensionFactory.new
  ef.subject_certificate = cert
  ef.issuer_certificate = issuer
  extensions.each do |oid, value, critical|
    cert.add_extension(ef.create_extension(oid, value, critical))
  end
  cert.sign(issuer_key, "sha256")
  cert
end
# rubocop:enable Naming/UncommunicativeMethodParamName


spec/statement_spec.rb
# frozen_string_literal: true

RSpec.describe SafetyNetAttestation::Statement do
  let(:root_key) { generate_key }
  let(:root_dn) { OpenSSL::X509::Name.new([["DC", "test"], ["CN", "Fake Google Root CA"]]) }
  let(:root_certificate) do
    extensions = [
      ["basicConstraints", "CA:TRUE", true],
      ["keyUsage", "keyCertSign,cRLSign", true]
    ]
    issue_cert(root_dn, root_key, 1, extensions)
  end

  let(:signing_key) { generate_key }
  let(:signing_dn) { OpenSSL::X509::Name.new([["CN", "attest.android.com"]]) }
  let(:signing_certificate) do
    extensions = [
      ["basicConstraints", "CA:FALSE", true],
      ["keyUsage", "digitalSignature,nonRepudiation", true]
    ]
    issue_cert(signing_dn, signing_key, 2, extensions, issuer: root_certificate, issuer_key: root_key)
  end

  let(:payload_timestamp) { Time.now.to_f - 5.0 }
  let(:payload) do
    {
      "timestampMs" => payload_timestamp * 1000.0,
      "nonce" => "R2Rra24fVm5xa2Mg",
      "apkPackageName" => "com.package.name.of.requesting.app",
      "apkCertificateDigestSha256" => [Digest::SHA256.base64digest("test")],
      "ctsProfileMatch" => true,
      "basicIntegrity" => true,
    }
  end
  let(:response) do
    JWT.encode(
      payload,
      signing_key,
      "RS256",
      x5c: [Base64.strict_encode64(signing_certificate.to_der)]
    )
  end

  let(:nonce) { "R2Rra24fVm5xa2Mg" }
  let(:leeway) { 60 }
  subject do
    described_class.new(response).verify(nonce, timestamp_leeway: leeway, trusted_certificates: [root_certificate])
  end

  it "returns itself and allows access to reader methods when everything is valid", :aggregate_failures do
    expect(subject).to be_a(described_class)

    expect(subject.cts_profile_match?).to be true
    expect(subject.basic_integrity?).to be true
    expect(subject.apk_package_name).to eq("com.package.name.of.requesting.app")
    expect(subject.apk_certificate_digest_sha256).to eq(["n4bQgYhMfWWaL+qgxVrQFaO/TxsrC4Is0V1sFbDwCgg="])
    expect(subject.error).to be_nil
    expect(subject.advice).to be_nil
    expect(subject.certificate_chain).not_to be_empty
    expect(subject.certificate_chain).to all(be_kind_of(OpenSSL::X509::Certificate))
  end

  it "loaded built-in certificates" do
    expect(described_class::GOOGLE_ROOT_CERTIFICATES).not_to be_empty
    expect(described_class::GOOGLE_ROOT_CERTIFICATES).to all(be_kind_of(OpenSSL::X509::Certificate))
  end

  context "using reader methods before validation" do
    subject { described_class.new(response) }

    it("is expected to raise SafetyNetAttestation::NotVerifiedError", :aggregate_failures) do
      expect { subject.cts_profile_match? }.to raise_error(SafetyNetAttestation::NotVerifiedError)
      expect { subject.basic_integrity? }.to raise_error(SafetyNetAttestation::NotVerifiedError)
      expect { subject.apk_package_name }.to raise_error(SafetyNetAttestation::NotVerifiedError)
      expect { subject.apk_certificate_digest_sha256 }.to raise_error(SafetyNetAttestation::NotVerifiedError)
      expect { subject.error }.to raise_error(SafetyNetAttestation::NotVerifiedError)
      expect { subject.advice }.to raise_error(SafetyNetAttestation::NotVerifiedError)
      expect { subject.certificate_chain }.to raise_error(SafetyNetAttestation::NotVerifiedError)
    end
  end

  context "invalid JWT" do
    context "wrong signing certificate subject" do
      let(:signing_dn) { OpenSSL::X509::Name.new([["CN", "foobar.android.com"]]) }

      it do
        expect { subject }.to raise_error(SafetyNetAttestation::CertificateSubjectError)
      end
    end

    context "wrong nonce" do
      let(:nonce) { "foobar0123456789" }

      it do
        expect { subject }.to raise_error(SafetyNetAttestation::NonceMismatchError)
      end
    end

    context "timestamp outside of leeway" do
      let(:payload_timestamp) { Time.now.to_f - (2 * leeway) }

      it do
        expect { subject }.to raise_error(SafetyNetAttestation::TimestampError)
      end
    end
  end

  context "example JWT" do
    let(:response) { File.read(File.join(__dir__, "webauthn-example-jwt.txt")) }
    let(:nonce) { "ywDhtBB5GEejNUbs2JrFKiU2RTlZPYXY3V4qBLYI5+c=" }
    let(:current_time) { Time.utc(2019, 7, 7, 16, 15, 11) }

    subject { described_class.new(response).verify(nonce, time: current_time) }

    it "returns itself and allows access to reader methods when everything is valid", :aggregate_failures do
      expect(subject).to be_a(described_class)

      expect(subject.cts_profile_match?).to be true
      expect(subject.basic_integrity?).to be true
      expect(subject.apk_package_name).to eq("com.google.android.gms")
      expect(subject.apk_certificate_digest_sha256).to eq(["8P1sW0EPJcslw7UzRsiXL64w+O50Ed+RBICtay1g24M="])
      expect(subject.error).to be_nil
      expect(subject.advice).to be_nil
      expect(subject.certificate_chain).not_to be_empty
      expect(subject.certificate_chain).to all(be_kind_of(OpenSSL::X509::Certificate))
    end
  end
end


spec/x5c_key_finder_spec.rb
# frozen_string_literal: true

require "spec_helper"
require "safety_net_attestation/x5c_key_finder"

RSpec.describe SafetyNetAttestation::X5cKeyFinder do
  let(:root_key) { generate_key }
  let(:root_dn) { OpenSSL::X509::Name.parse("/DC=org/DC=fake-ca/CN=Fake CA") }
  let(:root_certificate) do
    extensions = [
      ["basicConstraints", "CA:TRUE", true],
      ["keyUsage", "keyCertSign,cRLSign", true]
    ]
    issue_cert(root_dn, root_key, 1, extensions)
  end

  let(:leaf_key) { generate_key }
  let(:leaf_dn) { OpenSSL::X509::Name.parse("/DC=org/DC=fake/CN=Fake") }
  let(:leaf_serial) { 2 }
  let(:leaf_not_after) { Time.now + 3600 }
  let(:leaf_signing_key) { root_key }
  let(:leaf_certificate) do
    extensions = [
      ["basicConstraints", "CA:FALSE", true],
      ["keyUsage", "digitalSignature,nonRepudiation", true]
    ]
    issue_cert(leaf_dn, leaf_key, leaf_serial, extensions, issuer: root_certificate, issuer_key: leaf_signing_key,
                                                           not_after: leaf_not_after)
  end

  let(:x5c_certificates) { [leaf_certificate] }
  subject(:keyfinder) { described_class.from(x5c_certificates, [root_certificate]) }

  it "returns the certificate chain used to verify the JWT" do
    expect(keyfinder).to all(be_a(OpenSSL::X509::Certificate))
    expect(keyfinder.first.to_pem).to eq(leaf_certificate.to_pem)
    expect(keyfinder.last.to_pem).to eq(root_certificate.to_pem)
  end

  context "certificate" do
    context "expired" do
      let(:leaf_not_after) { Time.now - 3600 }

      it "raises an error" do
        error = "Certificate verification failed: certificate has expired. Certificate subject: " \
          "/DC=org/DC=fake/CN=Fake."
        expect { keyfinder }.to raise_error(SafetyNetAttestation::SignatureError, error)
      end
    end

    context "signature could not be verified with the given trusted roots" do
      let(:leaf_signing_key) { generate_key }

      it "raises an error" do
        error = "Certificate verification failed: certificate signature failure. Certificate subject: " \
          "/DC=org/DC=fake/CN=Fake."
        expect { keyfinder }.to raise_error(SafetyNetAttestation::SignatureError, error)
      end
    end

    context "could not be chained to a trusted root certificate" do
      subject(:keyfinder) { described_class.from(x5c_certificates, []) }

      it "raises an error" do
        error = "Certificate verification failed: unable to get local issuer certificate. Certificate subject: " \
          "/DC=org/DC=fake/CN=Fake."
        expect { keyfinder }.to raise_error(SafetyNetAttestation::SignatureError, error)
      end
    end
  end
end


lib/safety_net_attestation/x5c_key_finder.rb
# frozen_string_literal: true

require "base64"
require_relative "errors"

module SafetyNetAttestation
  class SignatureError < Error; end

  class X5cKeyFinder
    def self.from(x5c_certificates, trusted_certificates, time: Time.now)
      store = OpenSSL::X509::Store.new
      trusted_certificates.each { |certificate| store.add_cert(certificate) }

      signing_certificate, *certificate_chain = x5c_certificates
      store_context = OpenSSL::X509::StoreContext.new(store, signing_certificate, certificate_chain)
      store_context.time = time

      if store_context.verify
        store_context.chain
      else
        error = "Certificate verification failed: #{store_context.error_string}."
        error = "#{error} Certificate subject: #{store_context.current_cert.subject}." if store_context.current_cert

        raise SignatureError, error
      end
    end
  end
end


lib/safety_net_attestation/statement.rb
# frozen_string_literal: true

require "jwt"
require "openssl"
require "time"
require_relative "errors"
require_relative "fixed_length_secure_compare"
require_relative "x5c_key_finder"

module SafetyNetAttestation
  class Statement
    GOOGLE_ROOT_CERTIFICATES = Dir.glob(
      File.join(SafetyNetAttestation::GEM_ROOT, "safety_net_attestation", "certificates", "*.*")
    ).map do |path|
      file = File.binread(path)
      OpenSSL::X509::Certificate.new(file)
    end.freeze

    using FixedLengthSecureCompare

    attr_reader :json

    def initialize(jws_result)
      @jws_result = jws_result
    end

    def verify(nonce, timestamp_leeway: 60, trusted_certificates: GOOGLE_ROOT_CERTIFICATES, time: Time.now)
      certificate_chain = nil
      response, _ = JWT.decode(@jws_result, nil, true, algorithms: ["ES256", "RS256"]) do |headers|
        x5c_certificates = headers["x5c"].map do |encoded|
          OpenSSL::X509::Certificate.new(Base64.strict_decode64(encoded))
        end

        certificate_chain = X5cKeyFinder.from(x5c_certificates, trusted_certificates, time: time)
        certificate_chain.first.public_key
      end

      verify_certificate_subject(certificate_chain.first)
      verify_nonce(response, nonce)
      verify_timestamp(response, timestamp_leeway, time)

      @json = response
      @certificate_chain = certificate_chain
      self
    end

    def cts_profile_match?
      raise NotVerifiedError unless json

      json["ctsProfileMatch"]
    end

    def basic_integrity?
      raise NotVerifiedError unless json

      json["basicIntegrity"]
    end

    def apk_package_name
      raise NotVerifiedError unless json

      json["apkPackageName"]
    end

    def apk_certificate_digest_sha256
      raise NotVerifiedError unless json

      json["apkCertificateDigestSha256"]
    end

    def error
      raise NotVerifiedError unless json

      json["error"]
    end

    def advice
      raise NotVerifiedError unless json

      json["advice"]&.split(",")
    end

    def certificate_chain
      raise NotVerifiedError unless json

      @certificate_chain
    end

    private

    def verify_certificate_subject(certificate)
      common_name = certificate.subject.to_a.assoc("CN")

      unless common_name[1] == "attest.android.com"
        raise CertificateSubjectError
      end
    end

    def verify_nonce(response, nonce)
      unless OpenSSL.fixed_length_secure_compare(nonce, response["nonce"])
        raise NonceMismatchError
      end
    end

    def verify_timestamp(response, leeway, time)
      now = time.to_f
      response_time = response["timestampMs"] / 1000.0
      unless response_time.between?(now - leeway, now + leeway)
        raise TimestampError, "not within #{leeway}s leeway"
      end
    end
  end
end
